### PR TITLE
Let auth data be updated on login

### LINKED ISSUE #1455: "Facebook auth is invalid for this user." after trying to log in after app has been deleted.
When logging into our app we're using the PFFacebookUtils method `PFFacebookUtils.logInInBackgroundWithReadPermissions(self.permissions, block: {})` and that is working fine.  The first time that is ran, the callback has the `user.isNew` flag set to true and we go one path.

However I'm trying to test out the workflow where the user deletes the app from their phone and Facebook account.

In this instance when I hit the other path where `user.isNew` is false in the callback, the new time I go to save the user after making updates to it, I get the error `Facebook auth is invalid for this user.`.

Is this by design and is there something I have to do to get this workflow working?  Or is this a bug?

### LINKED ISSUE #1735: Facebook login is failing in Private parse cloud after Facebook password change
Check out https://github.com/ParsePlatform/parse-server/issues/1271 for an ideal bug report. The closer your issue report is to that one, the more likely we are to be able to help, and the more likely we will be to fix the issue quickly!

Many members of the community use Stack Overflow and Server Fault to ask questions. Read through the existing questions or ask your own!

- Stack Overflow: http://stackoverflow.com/questions/tagged/parse.com 
- Server Fault: https://serverfault.com/tags/parse

For database migration help, please file a bug report at https://parse.com/help#report

Make sure these boxes are checked before submitting your issue -- thanks for reporting issues back to Parse Server!

- [ ] You've met the prerequisites: https://github.com/ParsePlatform/parse-server/wiki/Parse-Server-Guide#prerequisites.

Yes

- [ ] You're running the latest version of Parse Server: https://github.com/ParsePlatform/parse-server/releases
Using  version 2.2.6

- [ ] You've searched through existing issues: https://github.com/ParsePlatform/Parse-Server/issues?utf8=%E2%9C%93&q=is%3Aissue Chances are that your issue has been reported or resolved before.

#### Environment Setup

- Server: parse-server version, operating system, hardware, local or remote?
- Database: version, storage engine, hardware, local or remote?
Using  version 2.2.6
#### Steps to reproduce
- I use PF Login to logging in IOS App.
- Login to IOS  app  using Facebook.
- Go to Facebook and change password
- logout of app 
-Login using Facebook again.

Alert comes up saying "Please log into this app again  to reconnect to your Facebook account."

After logging again same alert and I can't access my data linked to used id.

- Can this issue be reproduced using the Parse Server REST API? Include curl commands when applicable.
- What was the expected result? What is the actual outcome?

Didn't Try

#### Logs/Trace

- You can turn on additional logging by configuring VERBOSE=1 in your environment.

I have migrated to linux based Private parse environment in Rack space.
I use PF Login for user login using Facebook.

Here is Latest versions
Parse Server 2.2.6
FBSDKCoreKit 4.11.0
ParseUI (1.2.0)
ParseFacebookUtilsV4 (1.11.1)
IOS 8 and above

Everything have been working fine. Then I changed my Facebook password. When I am logging into app using Face book alert is coming up with message 

"Please log into this app again  to reconnect to your Facebook account."

I am logging out and login in again to get same message.

However other users who haven't changed password are able to login as usual.

To further check
I pointed my app to parse.com. It is working fine .

Any help is much appreciated.

### LINKED ISSUE #1947: Parse Server 2.2.10 not updating authData to latest on User save
### Issue Description

Parse Server 2.2.10 doesn't update authorization data when saved. There is some old access_token that keeps being assigned to my logged in user rather than the new one being provided. You can see in the following flow that I post a valid access token to the User class, and then I'm returned some other invalid access_token.
```
POST /parse/1/users { host: 'localhost:1337',
  'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:48.0) Gecko/20100101 Firefox/48.0',
  accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
  'accept-language': 'en-US,en;q=0.5',
  'accept-encoding': 'gzip, deflate',
  referer: 'http://localhost:63342/yorick/public/index.html',
  'content-type': 'text/plain',
  'content-length': '470',
  origin: 'http://localhost:63342',
  connection: 'keep-alive',
  pragma: 'no-cache',
  'cache-control': 'no-cache' } {
  "authData": {
    "facebook": {
      "id": "REDACTED",
      "access_token": "EAAW1sZB8mZAsQBAMD1SEGzLUioe3mKmvaHCI2VBXKAnyaa8qZAifu4Xfqk0FUVoLF0RZCYpHsXkDiOdLBZBpd2Ys9ujXkpNmhPrj4kGUunJv0w3QXHSagZAO6gzBAhfXyh2qq7NXzzjGhFxy4mg2vJUjerHcztYqLfdJin1HNpHgZDZD",
      "expiration_date": "2016-05-29T20:59:35.335Z"
    }
  }
}
verbose: {
  "response": {
    "ACL": {
      "*": {
        "read": true
      },
      "8JgU9bwIix": {
        "read": true,
        "write": true
      }
    },
    "objectId": "8JgU9bwIix",
    "username": "devuser",
    "updatedAt": "2016-05-29T19:13:16.871Z",
    "createdAt": "2016-05-27T07:10:42.724Z",
    "admininterface": false,
    "storytellerinterface": false,
    "authData": {
      "facebook": {
        "id": "REDACTED",
        "access_token": "EAAW1sZB8mZAsQBAGXJZA0kWDyJLBY2yDOIqTou9iBuRLpwOHgw2LgRj6MnAtMZBuar0QeQo0HYLPoZB6p9filfffMtBbgRTSiI0LpLcfbZASzLRE8M6nPhf9pKB6840aZCe3fjQCwPbJcRgMeGsYJB4sP2Vvdbuo3ZA2T7gG2OpvXgZDZD",
        "expiration_date": "2016-05-27T09:00:17.948Z"
      }
    },
    "email": "andrew.sayman@gmail.com",
    "realname": "Andrew Sayman",
    "sessionToken": "r:1e86f0b8d0697e4afd11d78c6718ad75"
  },
  "location": "http://localhost:1337/parse/1/users/8JgU9bwIix"
}
```
I can't actually figure out where the old value is even being stored. It doesn't seem to be in my database anywhere. I've deleted all of the sessions. On the user object associated with this facebook ID I even manually forced it to a bad value ("aoeu"), but this other value keeps coming back.

In both the database and the javascript session this bad access_token is stored. This means I can't save any updates to the User object because the check against this bad token fails.

[FILL THIS OUT]

### Steps to reproduce

1. Log in a user using Facebook
2. Fetch the current user object
3. Save the user object

#### Expected Results

User object is updated with new and valid data

#### Actual Outcome

Auth data is old and bad, so the object fails to save and I'm now not properly authenticated on the server side

### Environment Setup

- **Server**
  - parse-server version: 2.2.10
  - Operating System:     Windows 10, Debian 7.10, Ubuntu 14.04.4 LTS
  - Localhost or remote server? (AWS, Heroku, Azure, Digital Ocean, etc): Localhost, remote on Digital Ocean, remote on Nitrous.io

- **Database**
  - MongoDB version: 3.2
  - Storage engine:  Default
  - Localhost or remote server? (AWS, mLab, ObjectRocket, Digital Ocean, etc): Localhost, remote on Digital Ocean (Debian 7.10), localhost on Nitrous.io

### Logs/Trace
```
POST /parse/1/users { host: 'localhost:1337',
  'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:48.0) Gecko/20100101 Firefox/48.0',
  accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
  'accept-language': 'en-US,en;q=0.5',
  'accept-encoding': 'gzip, deflate',
  referer: 'http://localhost:63342/yorick/public/index.html',
  'content-type': 'text/plain',
  'content-length': '470',
  origin: 'http://localhost:63342',
  connection: 'keep-alive',
  pragma: 'no-cache',
  'cache-control': 'no-cache' } {
  "authData": {
    "facebook": {
      "id": "REDACTED",
      "access_token": "EAAW1sZB8mZAsQBAMD1SEGzLUioe3mKmvaHCI2VBXKAnyaa8qZAifu4Xfqk0FUVoLF0RZCYpHsXkDiOdLBZBpd2Ys9ujXkpNmhPrj4kGUunJv0w3QXHSagZAO6gzBAhfXyh2qq7NXzzjGhFxy4mg2vJUjerHcztYqLfdJin1HNpHgZDZD",
      "expiration_date": "2016-05-29T20:59:35.335Z"
    }
  }
}
verbose: {
  "response": {
    "ACL": {
      "*": {
        "read": true
      },
      "8JgU9bwIix": {
        "read": true,
        "write": true
      }
    },
    "objectId": "8JgU9bwIix",
    "username": "devuser",
    "updatedAt": "2016-05-29T19:13:16.871Z",
    "createdAt": "2016-05-27T07:10:42.724Z",
    "admininterface": false,
    "storytellerinterface": false,
    "authData": {
      "facebook": {
        "id": "REDACTED",
        "access_token": "EAAW1sZB8mZAsQBAGXJZA0kWDyJLBY2yDOIqTou9iBuRLpwOHgw2LgRj6MnAtMZBuar0QeQo0HYLPoZB6p9filfffMtBbgRTSiI0LpLcfbZASzLRE8M6nPhf9pKB6840aZCe3fjQCwPbJcRgMeGsYJB4sP2Vvdbuo3ZA2T7gG2OpvXgZDZD",
        "expiration_date": "2016-05-27T09:00:17.948Z"
      }
    },
    "email": "andrew.sayman@gmail.com",
    "realname": "Andrew Sayman",
    "sessionToken": "r:1e86f0b8d0697e4afd11d78c6718ad75"
  },
  "location": "http://localhost:1337/parse/1/users/8JgU9bwIix"
}
```

### PR DESCRIPTION
**Pull request by flovilmart**
This proposal will let user auth data to be updated when a valid token is provided with a different expiration time for the same authData.id.

Fixes #1947
Fixes #1735
Fixes #1455

Probably fixes other issues as well.

### CODE REVIEW

File: spec/ParseUser.spec.js [hunk 1]
```diff
@@ -1428,6 +1466,47 @@ describe('Parse.User testing', () => {
     });
   });
 
+  fit_exclude_dbs(['postgres'])("link multiple providers and updates token", (done) => {
```
Comment (steven-supersolid) [inline spec/ParseUser.spec.js]: Probably want to change from fit to it here as I think this will force a single test

Comment (flovilmart) [inline spec/ParseUser.spec.js]: OOoopppssss

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
thanks @steven-supersolid !

**Comment by mfpiccolo** [comment]
👍 Thanks for getting this.  We are waiting for this to go through.

**Comment by flovilmart** [comment]
Waiting for a formal review approval before merging.

**Comment by acegreen** [comment]
@flovilmart ah come on people just merge it. This is critical.

**Comment by flovilmart** [comment]
That's why I'm waiting for a proper review...

**Comment by acegreen** [comment]
@flovilmart forget the proper review, you seem like a smart guy..ahaha

**Comment by towen** [comment]
@flovilmart Great job on getting this in. +1 on getting this merged soon. This is the only hurdle I have on getting a new client launched...

**Comment by acegreen** [comment]
@flovilmart thanks for the great work! Is his available through the npm install? I see that the parse example still uses 2.2.12

**Comment by acegreen** [comment]
A related issue??

https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/issues/954

**Comment by flovilmart** [comment]
Not sure it's related...

**Comment by vandutran** [comment]
Will this fix this issue?

https://github.com/ParsePlatform/ParseFacebookUtils-iOS/issues/72

**Comment by flovilmart** [comment]
It should have already as it's been merged a month ago :)

**Pull request merged**